artemis-commons/src/main/java/org/apache/activemq/artemis/utils/DefaultSensitiveStringCodec.java

```diff
@@ -188,6 +208,16 @@ public String encode(String secret) throws Exception {
          BigInteger n = new BigInteger(encoding);
          return n.toString(16);
       }
+
+      @Override
+      public boolean verify(char[] inputValue, String storedValue) {
+         try {
+            return Objects.equals(storedValue, encode(String.valueOf(inputValue)));
```

Comment:
The methods here seems somewhat all over the place.

The class implements SensitiveDataCodec<String>, meaning it has to return String (i.e T) for encode and decode methods, but is only taking Object args which it then blindly assumes are actually String and casts to that unchecked.

The new compare method also takes Object args and then for some odd reason requires one of them to be String or char[], but only String in the other. It converts the first to char[] if it wasnt, so it can call this verify method which has fixed char[] and String type args. Which then turns the char[] _[back]_ into a String so it can call encode with a String to get another String to further compare to the given String.

Why allow different types in the various different methods? Why is one methods args fixed type and the others floating Object, that are then assumed or enforced to be a specific types (meaning the broker has to be calling it with those types), and yet always ending up as String ultimately. Almost nothing is using the generic T type to pick up the String generic from the class definition but that appears to be the only type that is used in the broker so nothing else will actually work.

Actually, why add both new verify and compare methods when they do essentially the same thing overall? Verify is probably a better name overall since compare doesnt as much imply the fact one arg has to be manipulated to equal the other (and the not-present javadoc doesnt say what the method actually does at all)

Reply:
The compare method has Object args to match other method signatures and to compatible with the existing verify method of DefaultSensitiveStringCodec. See my other comment on SecureHashProcessor for further details.

Reply:
Match/compatible in what way though? The existing encode/decode methods args are blindly assumed _unchecked_ to have to be String. The broker only expects and provides String objects for those method, with the overall type used also hard coded to require a SensitiveDataCodec\<String\>.

For SecureHashProcessor compare(), it is [_still_] fixed to char[] + String, as is DefaultSensitiveStringCodec verify (and its internal CodecAlgorithm, which doesnt actually use Object for any of the methods). It seems that adding e.g SensitiveDataCodec.verify(char[], T) would be nicer and/or simpler, easily get rid of a bunch of duplication and unnecessary arg validation. A verify(T, T) could be added compatibly later if actually needed.

Reply:
Looking existing encode/decode methods of SensitiveDataCodec<T> interface with Object args, I think they were deigned to encode/decode different sensitive object types to type T. So IMO the type T is to define the outputs not the inputs in the original design. I used the Object args for the compare method to keep the original design of the interface.

While SensitiveDataCodec.verify(char[], T) is fine if T is String, it could not work well when T is not String.

Reply:
I agree on SensitiveDataCodec.verify(T, T), I was against SensitiveDataCodec.verify(char[], T)

Reply:
I'm fine with that also, though I think passing it a char[] is perfectly reasonable given the use case. Just not defining entirely unnecessary Object args that only muddy the waters given the usage.

Reply:
Used SensitiveDataCodec.verify(T, T)